[flang] AliasAnalysis: Fix pointer component logic

jdenny-ornl opened the pull request:
This PR applies the changes discussed in [[RFC] Rationale for Flang AliasAnalysis pointer component logic](https://discourse.llvm.org/t/rfc-rationale-for-flang-aliasanalysis-pointer-component-logic/79252).

In summary, this PR replaces the existing pointer component logic in Flang's AliasAnalysis implementation.  That logic focuses on aliasing between pointers and non-pointer, non-target composites that have pointer components.  However, it is more conservative than necessary, and some existing tests expect its current results when less conservative results seem reasonable.

This PR splits the logic into two cases:

1. Source values are the same: Return MayAlias when one value is the address of a composite, and the other value is statically the address of a pointer component of that composite.
2. Source values are different: Return MayAlias when one value is the address of a composite (actual argument), and the other value is the address of a pointer (dummy arg) that might dynamically be a component of that composite.

In both cases, the actual implementation is still more conservative than described above, but it can be improved further later.  Details appear in the comments.

Additionally, this PR revises the logic that reports MayAlias for a pointer/target vs. another pointer/target.  It constrains the existing logic to handle only isData cases, and it adds less conservative handling of !isData cases elsewhere.  First, it extends case 2 listed above to cover the case where the actual argument is the address of a pointer rather than a composite.  Second, it adds a third case: where target attributes enable aliasing with a dummy argument.

jdenny-ornl committed:
WIP: [flang] AliasAnalysis: Fix pointer component logic

jdenny-ornl committed:
WIP: Fix case of ptr dummy arg vs. ptr component

jdenny-ornl commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -122,13 +122,32 @@ AliasResult AliasAnalysis::alias(Value lhs, Value rhs) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "  aliasing between composite and pointer component\n");
+      return AliasResult::MayAlias;
+    }
```
@Renaud-K posted the following [comment as part of a commit](https://github.com/jdenny-ornl/llvm-project/commit/484cb901cb3f5ae934b55629bb8537e42d3a832d#r142575988), and I'm trying to move the conversation into this PR:

> I don't think `approximateSource` is a reliable way of meaning `component`. We could have an optimized out `fir.ccordinate_of %memref %c0`, still referring to the same source.

jdenny-ornl replied:
@Renaud-K I could extend `Source` with a field like `bool component;` It would be set to true for any `DesignateOp` specifying a component, and `approximateSource` would also still be set to true then.  Does that seem like a reasonable direction to try?

Renaud-K replied:
`DesignateOp` is later lowered to `coordinate_of` for instance. The absence of which does not mean we are not dealing with a component. The alias analysis has to work at all levels.  Again, `fir.coordinate_of %memref %c0` is the same as `%memref`. We need to provide consistent results whether `fir.coordinate_of` is present or at least error on the conservative side.

jdenny-ornl replied:
> `DesignateOp` is later lowered to `coordinate_of` for instance.

Thanks.  In a simple example, I see that

```
%10 = hlfir.designate %1#0{"p"}   {fortran_attrs = #fir.var_attrs<pointer>} : (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<f32>>}>>) -> !fir.ref<!fir.box<!fir.ptr<f32>>>
```

becomes

```
%10 = fir.field_index p, !fir.type<_QMmTt{p:!fir.box<!fir.ptr<f32>>}>
%11 = fir.coordinate_of %1, %10 : (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<f32>>}>>, !fir.field) -> !fir.ref<!fir.box<!fir.ptr<f32>>>
```

The latter pattern seems simple enough to detect.  I did not find a later version of that IR where `fir.coordinate_of` appeared but `!fir.field` and `fir.field_index` had been optimized away.  Is your `fir.coordinate_of %memref %c0` an example of that?

> The absence of which does not mean we are not dealing with a component.

Is it true that either `hlfir.designate` or `fir.coordinate_of` must be present when accessing a component?  If not, then, independently of this PR, we risk `approximateSource=false` and an invalid MustAlias, right?

Renaud-K replied:
MustAlias is correct if we can determine the source more accurately, and MayAlias is a conservative error. But in your case we could go from MayAlias (because we detect components) to NoAlias which is too optimistic. Maybe we can keep using the type, such as, if the type of the source is a derived type and the type of the starting point is not then we have a component.

jdenny-ornl replied:
This is almost what you suggested:

```
     if (lhsSrc.origin.u == rhsSrc.origin.u &&
-        ((lhsSrc.approximateSource && !lhsSrc.isData() && !rhsSrc.approximateSource) ||
-         (rhsSrc.approximateSource && !rhsSrc.isData() && !lhsSrc.approximateSource))) {
+        ((lhsSrc.isRecordWithPointerComponent() && !rhsSrc.isData()) ||
+         (rhsSrc.isRecordWithPointerComponent() && !lhsSrc.isData()))) {
       LLVM_DEBUG(llvm::dbgs()
                  << "  aliasing between composite and non-data component\n");
       return AliasResult::MayAlias;
     }
```

Renaud-K replied:
Yes. Thank you. Now from a testing stand point I am not sure if we explicitly test for the case of component pointers **data**. Maybe they will be handled by the general processing of pointers and targets but it would be nice to check for aliasing since they should be aliasing with other target or pointers. Do you agree with this statement?

jdenny-ornl replied:
> Now from a testing stand point I am not sure if we explicitly test for the case of component pointers **data**.

That requires a load to extract the address in the pointer.  That's the address of the data.  However, getSource doesn't handle load on, for example, an hlfir.designate yet, so we get an Indirect and thus a MayAlias.  Is that what you mean?

I am working on a patch to extend load handling to that case.  The general pointer component logic confused my testing efforts, so I wanted to address it first.

Renaud-K replied:
Hmm, we do handle loads as long as they are not dynamic. If we cannot statically determine what is loaded (the comment calls it a leaf source), then it is indirect. There is not much we can do about it. What is not yet handled is the case when the leaf source is a SourceKind::Allocate but that should not prevent you from working with pointers and dummy arguments.

jdenny-ornl replied:
%11 below is the address of the data in `x%p` where `p` is a pointer:

```
  func.func @_QMmPfoo() {
    ...
    %2 = fir.alloca !fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}> {bindc_name = "x", uniq_name = "_QMmFfooEx"}
    %3:2 = hlfir.declare %2 {uniq_name = "_QMmFfooEx"} : (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}>>) -> (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}>>, !fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}>>)
    ...
    %9 = hlfir.designate %3#0{"p"}   {fortran_attrs = #fir.var_attrs<pointer>} : (!fir.ref<!fir.type<_QMmTt{p:!fir.box<!fir.ptr<i32>>}>>) -> !fir.ref<!fir.box<!fir.ptr<i32>>>
    %10 = fir.load %9 : !fir.ref<!fir.box<!fir.ptr<i32>>>
    %11 = fir.box_addr %10 : (!fir.box<!fir.ptr<i32>>) -> !fir.ptr<i32>
    ...
    return
  }
```

The LoadOp case in getSource needs to be extended to handle the hlfir.designate.  My current draft of that patch does it in getOriginalDef.

I also have a patch to extend LoadOp for alloca.

Renaud-K replied:
I am not sure we can do anything about `hlfir.designate` in `getOriginalDef`. The value loaded from %9 could be anything. We cannot follow it to %3. A load from %3 has nothing to do with a load from %9.

jdenny-ornl replied:
%11 below is the address of the data in p, which is a pointer:

```
  func.func @_QMmPfoo() {
    ...
    %2 = fir.alloca !fir.box<!fir.ptr<i32>> {bindc_name = "p", uniq_name = "_QMmFfooEp"}
    %3:2 = hlfir.declare %2 {fortran_attrs = #fir.var_attrs<pointer>, uniq_name = "_QMmFfooEp"} : (!fir.ref<!fir.box<!fir.ptr<i32>>>) -> (!fir.ref<!fir.box<!fir.ptr<i32>>>, !fir.ref<!fir.box<!fir.ptr<i32>>>)
   ...
    %10 = fir.load %3#0 : !fir.ref<!fir.box<!fir.ptr<i32>>>
    %11 = fir.box_addr %10 : (!fir.box<!fir.ptr<i32>>) -> !fir.ptr<i32>
    ...
    return
  }
```

Isn't this the same as the previous example except that, without the designate, we're loading all of %2 instead of only one component of it?

jdenny-ornl replied:
To finish the thought, I don't see why we have to give up (SourceKind=Indirect forces MayAlias) on tracking aliasing for a pointer just because it's a member of an alloca rather than the entire alloca.

Renaud-K replied:
My concern would be to modify getOriginalDef to return anything other than Source::Indirect. 
`%11` with the designate case is always indirect. We have no idea what was stored in `%9`, it could a global, an allocatable or a dummy: we do not know.
 
There is room to improve on SourceKind::Indirect, such as using the type system to distinguish !fir.ptr<T> from other memory references but that would be a different discussion.  In the example just above, `%11` would then be indirect too since we are following the data and not interested in the box address which happens to be an SourceKind::Alloca. Today, I would guess, we would get SourceKind::Alloca with an isData flag but it would be better modeled with SourceKind::Indirect and isData. Then we could do with SourceKind::Indirect what we used to do clumsily with SourceKind::Direct. 

But I think, we should take it one step at a time.

jdenny-ornl replied:
> My concern would be to modify getOriginalDef to return anything other than Source::Indirect.

For the record, getOriginalDef does not itself choose a SourceKind.  Moreover, after getSource's LoadOp case calls getOriginalDef, getSource currently returns any of Indirect, Argument, or Global.  That is true independently of this patch.  No modification required.

(As you have noted, there are cases like alloca that getSource's LoadOp case doesn't currently fully handle.  As a result, it currently returns Indirect sometimes where, for consistency with the current handling of dummy args and globals, it should return Allocate.)

> `%11` with the designate case is always indirect. We have no idea what was stored in `%9`, it could a global, an allocatable or a dummy: we do not know.

In that example, what was stored in %9 is the address of the pointer. In my above example without hlfir.designate, %3 is the address of the pointer.  As far as I understand, the difference between those is whether that address is at a base address (for a global, dummy arg, alloca, etc.) or at an offset from that base address.

I don't see why that difference *should* affect the SourceKind for any value that is computed from the address of the pointer.  In particular, if the source origin (i.e., %2) were a dummy arg or global, the SourceKind for %11 in my above example without hflir.designate would currently be Argument or Global, but the SourceKind for %11 in my above example with hlfir.designate would currently be Indirect.  I see no reason for that difference.  In a later patch, I would like to fix that.  (Again, I would also like to add support for the alloca/Allocate case shown in the examples.)

> There is room to improve on SourceKind::Indirect, such as using the type system to distinguish !fir.ptr from other memory references but that would be a different discussion. In the example just above, `%11` would then be indirect too since we are following the data and not interested in the box address which happens to be an SourceKind::Alloca. Today, I would guess, we would get SourceKind::Alloca with an isData flag but it would be better modeled with SourceKind::Indirect and isData. Then we could do with SourceKind::Indirect what we used to do clumsily with SourceKind::Direct.

Whatever we decide to call the SourceKind for that case, do we agree that hlfir.designate should not change the SourceKind for %11 between my previous two examples?

> But I think, we should take it one step at a time.

Agreed.  Hopefully the current patch I'm trying to develop can be a next step.

> It would impact @tblah who is handling the (global, isData) separately. These two would become (indirect, isData)

Where is the discussion of his work?

tblah replied:
@jdenny-ornl my work was proposed here https://discourse.llvm.org/t/rfc-propagate-fir-alias-analysis-information-using-tbaa/73755

There have since been changes since then to add support for Direct variables: https://github.com/llvm/llvm-project/commit/6242c8ca18bcd6765094c73dd2c8b49200a6cec8

I think in this case the proposed changes won't alter the TBAA tags. My pass doesn't tag SourceKind::Alloca (by default) or SourceKind::Indirect (at all) (there are still TBAA tags added later during codegen to distinguish box metadata access from data access, but that is different again).

jdenny-ornl replied:
@tblah Thanks for the pointers!

jdenny-ornl committed:
Fix logic in various ways, and add some tests

For the static case (when source values are the same):
- As discussed in this PR with Renaud Kauffmann, rely on types and
  `!isData` instead of `approximateSource`, which doesn't necessarily
  indicate a component.
- Improve comments to document effect on allocatables and some
  remaining limitations.

For the dynamic case:
- For the composite, as discussed in the RFC, don't restrict it to a
  global.  It might be an argument.  Just exclude locals, which cannot
  be aliased by a pointer dummy arg.
- For the pointer, make sure it's the address of the pointer
  (`!isData`) not the address in the pointer.
- For the pointer, make sure it's a dummy argument not a component of
  a composite dummy argument.

For both cases, for the composite, call `isRecordWithPointerComponent`
on the value (`lhs` or `rhs`) not on its source.  Otherwise, we can
mistake a separate component for a composite.

Fix `getSource` to record pointer attributes for `DesignateOp`.
Otherwise, we miss MayAlias for pointer components vs. other pointers.
Before the changes in this PR, `isRecordWithPointerComponent` returned
true on the source of a pointer component, and that was enough to
return MayAlias for it vs. another pointer.

Extend tests, including coverage of all fixes above.

Renaud-K commented on flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir:
```diff
@@ -17,13 +25,13 @@
 // end module
 
```
Can we check aliasing between x%next and and y%next?

jdenny-ornl replied:
Thanks for reviewing.  I've added that check.  It returns MayAlias.  As the comment I added says, I think it should return NoAlias.  We could try to fix that in this patch, but I think it would be better to address it in a separate patch because it involves a different part of the AliasAnalysis logic: the isTargetOrPointer() checks.

Renaud-K replied:
I pushed the cases I wanted covered here: https://github.com/llvm/llvm-project/pull/95287

jdenny-ornl replied:
Thanks.  I plan to take a look today.

jdenny-ornl replied:
That part looks good to me.  Thanks.

jdenny-ornl replied:
> Thanks for reviewing. I've added that check. It returns MayAlias. As the comment I added says, I think it should return NoAlias. We could try to fix that in this patch, but I think it would be better to address it in a separate patch because it involves a different part of the AliasAnalysis logic: the isTargetOrPointer() checks.

I decided to go ahead and [fix that](https://github.com/llvm/llvm-project/pull/94242/commits/ee248f5236ec4914cf8708167fd74f4b48c2733f) in this PR as it helped me to work through another issue.

jdenny-ornl committed:
Merge commit 'b3b9f8dd4cad' into flang-aa-ptr-component

jdenny-ornl committed:
Check x%next vs. y%next

vzakhari commented on flang/include/flang/Optimizer/Analysis/AliasAnalysis.h:
```diff
@@ -153,17 +153,16 @@ struct AliasAnalysis {
 
+    /// Return true, if `ty` is a reference type to an object of derived type
+    /// that contains a component with POINTER attribute.
+    static bool isRecordWithPointerComponent(mlir::Type ty);
```
nit: it may be better to make it a private static member of the `AliasAnalysis` class.

jdenny-ornl replied:
Same for the `isPointerReference` member function that follows it?

jdenny-ornl replied:
Done. Thanks for the review.

vzakhari reviewed (approved):
Thank you for the changes!

jdenny-ornl committed:
Merge branch 'main' into flang-aa-ptr-component

jdenny-ornl committed:
Drop accidental -debug from fir-opt in test

However, FileCheck directives then fail.  fir-opt's stdout (MLIR
output) and stderr (diagnostics) are jumbled together, and somehow
-debug changes the jumble sufficiently for FileCheck directives to
pass.  We don't actually need stdout, so send it to /dev/null.  That
makes the output easier to read anyway when debugging.

jdenny-ornl committed:
Relocate some functions as requested by reviewer

jdenny-ornl commented:
I plan to land this tomorrow.  Please let me know if anyone needs more time to review.

jeanPerier commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -159,21 +193,56 @@ AliasResult AliasAnalysis::alias(Value lhs, Value rhs) {
+       src1->kind != SourceKind::Allocate &&
+       src2->kind == SourceKind::Argument &&
+       src2->attributes.test(Attribute::Pointer) && !src2->isData() &&
+       !isRecordWithPointerComponent(src2->valueType)) ||
```
Why is the test "!isRecordWithPointerComponent(src2->valueType)" needed?
It seems either wrong or useless to me (the pointer component could itself point towards a derived type with pointer component).

jdenny-ornl replied:
Thanks for reviewing.  

The `if` preceding this one uses `isTargetOrPointer` and covers aliasing among addresses *in* pointers (what they point towards).  The `if` here has the `!src2->isData()` condition and so examines only the address *of* the pointer.

Specifically, the `if` here checks for aliasing with the address of a pointer that is a dummy argument.  The condition `!isRecordWithPointerComponent(src2->valueType)` prevents it from mistaking a pointer component of a dummy argument as a pointer dummy argument.  For example, [the test alias-analysis-9.fir has a check](https://github.com/jdenny-ornl/llvm-project/blob/1e5fdd49f4c8558f5181b3bdb2d38002e0b34630/flang/test/Analysis/AliasAnalysis/alias-analysis-9.fir#L31-L34) whose result would change without that condition:

```
// module m
// type t
//  type(t), pointer :: next
//  integer :: i
// end type
// contains
// subroutine foo(x, y)
//   type(t) :: x, y
//   integer :: i1, i2
//   i1 = x%next%i
//   x = y
//   i2 = x%next%i
// end subroutine
// end module

// CHECK-DAG: y#0 <-> xnext1#0: NoAlias
```

`y#0` is the address of `y`.  `xnext1#0` is the address of `x%next`. Without the above condition, AliasAnalysis would treat `xnext1#0` as the address of a pointer dummy argument, and the result above would become MayAlias.

jeanPerier replied:
Thanks for the explanation, while I understand the goal, the condition is still wrong in my opinion.

> the result above would become MayAlias

That would be a valid if the dummies had the TARGET attribute as per 15.5.2.14 (3) (b).

Here is an illustration where "a" and "c%p" should be considered to potentially alias (which is not enforced properly because of this condition). :
```
module m
  type t 
     real, pointer :: p
  end type
  type(t) :: a
  type(t) :: b
contains
subroutine test(c)
  type(t), target :: c
  c%p = 42
  a = b
  print *, c%p
end subroutine 
end module
    
  use m
  real, target :: x1 = 1
  real, target :: x2 = 2
  a%p => x1
  b%p => x2
  call test(a)
end 
```

To pass this at the FIR level, the `src2->attributes.test(Attribute::Pointer)` should also be removed because the attributes where not gathered on the hlfir.designate which does not exist anymore in FIR (replaced by fir.coordinate). This makes some sense anyway since the source (the dummy) in itself does not really has that attribute anyway.

jdenny-ornl replied:
I find it confusing that the `target` attribute on `c` is relevant in your example given that there are no pointers or other targets with the type of `c`.

Here is Fortran 2023 sec. 15.5.2.14, 1 (3) (b):

> 1 While an entity is associated with a dummy argument, the following restrictions hold.
>
> (3) Action that affects the value of the entity or any subobject of it shall be taken only through the dummy argument unless
>
> (b) the dummy argument is a scalar, assumed-shape, or assumed-rank object, and has the TARGET attribute but not the INTENT (IN) or CONTIGUOUS attributes, and the actual argument is a target other than a coindexed object or an array section with a vector subscript,

`c` is the dummy argument.  `a` is the actual argument.  But the above requires that "the actual argument is a target", which 3.141 defines as:

> entity that is pointer associated with a pointer (3.104), entity on the right-hand-side of a pointer assignment statement, or entity with the TARGET attribute

I don't see how `a` is a target in your example, so I don't see how your example is valid Fortran.  I think we can fix your example by adding the `target` attribute to `a`.  Do you agree?

It's a surprise to me to that the `target` attribute can be used to permit aliasing with other targets via dummy arguments.  I thought the sole purpose of the `target` attribute was to enable an entity to be the target of a pointer.

I want to make sure we're on the same page so far before I start working on fixing this case.

jeanPerier replied:
`a` in my example does not need target because it is not a dummy argument, so `15.5.2.14` does not restrict the effects on it. `15.5.2.14` restrict the effects on `c`, so it only matters that `c` has this attribute to allow `c` to be effected by changes on `a`.

I agree this is not a very well known aspect, but it is what is, and other compiler respect it (if you do A= B with two TARGET assumed-shape, you should observe that all compiler will cope with the case where A an B alias, even f not POINTERs are involved).

jdenny-ornl replied:
> `a` in my example does not need target because it is not a dummy argument, so `15.5.2.14` does not restrict the effects on it. `15.5.2.14` restrict the effects on `c`, so it only matters that `c` has this attribute to allow `c` to be effected by changes on `a`.

Thanks, but would you please help me to understand how the spec says that?  Or maybe the spec is stricter than defacto behavior?  In particular, when the dummy argument (`c`) has the TARGET attribute, one condition in the spec to permit aliasing is that the _actual argument_ (`a`) is a target.  Given the definition of "target" that I quoted, how is `a` a target in your example?
> 
> I agree this is not a very well known aspect, but it is what is, and other compiler respect it (if you do A= B with two TARGET assumed-shape, you should observe that all compiler will cope with the case where A an B alias, even f not POINTERs are involved).

I assume this doesn't refer to the `a` and `b` in your example given that neither have TARGET.  In this new example, are A and B both dummy arguments, both with TARGET?

jeanPerier replied:
Thank you for correcting me here, yes TARGET is needed on `a` for my example to be valid after reading that again. I verified with a few compilers that the behavior in `A=B` differs depending on A attribute in the following program (prints 1.0 if A has the attribute, prints N otherwise):

```
module m
integer, parameter :: n = 10000 
real, target :: a(n) = [(i, i=1,n)]
contains

subroutine test_aliasing(b)
  real, target :: b(:)
  a = b
end subroutine
end module

  use m
  call test_aliasing(a(n:1:-1))
  print *, a(n)
end
```

And yes, that was different A and B, sorry for the confusion.

jdenny-ornl replied:
Thanks for confirming.  I'll work on a fix.

jdenny-ornl replied:
I [added a fix](https://github.com/llvm/llvm-project/pull/94242/commits/a61be3a0d7202da4ccd9a3b38dc68dc3f119c2ce) for that.

jeanPerier commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -131,12 +160,17 @@ AliasResult AliasAnalysis::alias(Value lhs, Value rhs) {
   }
 
   Source *src1, *src2;
+  Value *val1, *val2;
```
nit: mlir::Value (mlir:: namespace is always spelled out in flang tree).

jdenny-ornl replied:
Should I also change the parameters of `AliasAnalysis::alias`?

jeanPerier replied:
Yes please.

jdenny-ornl replied:
Done.

jeanPerier commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -60,7 +60,15 @@ void AliasAnalysis::Source::print(llvm::raw_ostream &os) const {
 
-bool AliasAnalysis::Source::isPointerReference(mlir::Type ty) {
+bool AliasAnalysis::isRecordWithPointerComponent(mlir::Type ty) {
+  auto eleTy = fir::dyn_cast_ptrEleTy(ty);
```
Not directly related to your patch (since you only moved that code), but I found it while testing the logic below.

This should actually be `fir::unwrapSequenceType(dyn_cast_ptrOrBoxEleTy(ty))` in order to also unwrap fir.box type and arrays.

Otherwise, this will miss cases as the one below where the descriptor address of "p" may alias with the data of "c".

```
module m                                                           
  type t                                                           
     real, pointer :: p                                                                
  end type                                                            
  type(t) :: a(1)                                                           
  type(t) :: b(1)                                                           
contains                                                            
subroutine test(p, c)                                                           
  real, pointer :: p                                                           
  type(t), target :: c(:)                                                              
  p = 42                                                            
  c = b                                                           
  print *, p                                                           
end subroutine                                                            
end module                                                            
                                                            
  use m                                                           
  real, target :: x1 = 1                                                               
  real, target :: x2 = 2                                                               
  a(1)%p => x1                                                           
  b(1)%p => x2                                                           
  call test(a(1)%p, a)                                                                 
end  
```

jdenny-ornl replied:
Thanks for pointing that out.  After some of the other issues settle, I'll think about applying that, perhaps in another PR.

jdenny-ornl committed:
Value -> mlir::Value

jdenny-ornl commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -159,21 +193,56 @@ AliasResult AliasAnalysis::alias(Value lhs, Value rhs) {
+  // association during the assignment a = b.  Thus, the program should print 2.
+  if ((isRecordWithPointerComponent(val1->getType()) &&
+       src1->kind != SourceKind::Allocate &&
+       src2->kind == SourceKind::Argument &&
```
Make sense.  Thanks for catching that.  I'll work on adding a fix and test.

jdenny-ornl replied:
Done.

jdenny-ornl committed:
Fix and test HostAssoc case

As discussed in review.

jdenny-ornl committed:
Fix and test after convert-hlfir-to-fir

In particular, after hlfir.designate becomes fir.coordinate_of.

jdenny-ornl commented:
To avoid confusing [the thread on target attributes](https://github.com/llvm/llvm-project/pull/94242#discussion_r1645883337), I'm attempting to start a new one on one of the issues raised there:

> attributes where not gathered on the hlfir.designate which does not exist anymore in FIR (replaced by fir.coordinate)

I just pushed a commit to fix and test this issue.

jeanPerier commented:
> To avoid confusing [the thread on target attributes](https://github.com/llvm/llvm-project/pull/94242#discussion_r1645883337), I'm attempting to start a new one on one of the issues raised there:
> 
> > attributes where not gathered on the hlfir.designate which does not exist anymore in FIR (replaced by fir.coordinate)
> 
> I just pushed a commit to fix and test this issue.

I am a bit wary of the approach here. I do not like that we have to do work to get correctness (we should do work to prove something is not a pointer/target as much as possible). I see easy extensibility bugs when new ops are added. I would rather an approach where we have a mayBeATargetOrPointer flag set to true by default and that we set to false when we can prove it.

This could be thought in a different patch, and the current one could just be a bit more pessimistic and assume all descriptor addresses are for pointers.

jdenny-ornl commented:
> I am a bit wary of the approach here. I do not like that we have to do work to get correctness (we should do work to prove something is not a pointer/target as much as possible). I see easy extensibility bugs when new ops are added.

I understand your point, but how far would we take such an approach?  I copied the solution for determining `Attribute::Pointer` on `fir::CoordinateOp` from the existing solution on both `fir::AddrOfOp` and dummy args.  Is there something unique about `fir::CoordinateOp` that gives you pause?  Without this PR, do you find existing uses of `Source::isTargetOrPointer` problematic?

>  I would rather an approach where we have a mayBeATargetOrPointer flag set to true by default and that we set to false when we can prove it.

Would it maintain a list of cases (e.g., dummy args, `fir::AddrOfOp`, and possibly fir::CoordinateOp` after my solution) where we are confident we can determine no pointer/target?

jdenny-ornl committed:
Merge branch 'main' into flang-aa-ptr-component

jdenny-ornl committed:
Encapsulate repeated logic into functions

jdenny-ornl committed:
Fix todos for handling addresses of pointers

The general `if` calling `isTargetOrPointer` was handling addresses
*of* pointers similarly to addresses *in* pointers: it assumed they
may alias each other.  While that's true in the case of dummy args,
it's not true generally, as revealed by some of the test suite todos
fixed by this commit.

jdenny-ornl committed:
Fix handling of dummy arg with target attribute

That is, the address of a dummy arg may alias the address of a global
or another dummy arg when both have target attributes.  If either is a
composite, components may alias as well.

However, this PR broke handling of that case in two previous changes:

1. Initial commits in this PR replaced the conservative handling of
   pointer components.  As a result, they stopped reporting MayAlias
   for, e.g., the address of a pointer component of a dummy arg
   composite with a target attribute vs. the address of another
   composite with a target attribute.  This problem was discussed in
   the PR:
   <https://github.com/llvm/llvm-project/pull/94242#discussion_r1645883337>

2. The previous commit in this PR changed the conservative
   targets/pointers handling not to handle the case where both values
   have `!isData()`.  As a result, it stopped reporting MayAlias for,
   e.g., the address of an allocatable dummy arg with a target
   attribute vs. another allocatable with a target attribute.

This commit fixes such cases.

jdenny-ornl committed:
Give test more meaningful name

jdenny-ornl committed:
Merge branch 'main' into flang-aa-ptr-component

jdenny-ornl commented:
Ping.  @jeanPerier

clementval commented:
> Ping. @jeanPerier

FYI Jean is currently in PTO. I think he will be back next week.

jeanPerier reviewed (commented):
Thanks @jdenny-ornl for the update. This looks better!

I am still a bit bothered by the fact that the code is turned in a way were it is proving aliasing instead of no aliasing (i.e., it is doing  `if (cdt1) then mayalias else if (cdt2) then mayalias .... else noalias` ), this makes it harder to validate that aliasing cases are not missing via human review (at least in my experience). But I have to concede the standard is a bit written in that way in section 15.5.2.14 of F2023, and I do not have immediate suggestions to reverse the logic in a clearer way.

Can I ask you to push the `isRecordWithPointerComponent` and `isPointerReference` NFC change as well as the alias-analysis-9.fir renaming in a preliminary patch so that your functional changes stands out?

Then I will run this update through some end-to-end apps and keep you posted.

Do you know any apps where I should see performance improvement (i.e., was your change motivated by some use case in some app?). Note that in general I think we should try to change aliasing code in the future either to simplify the logic, or when motivated by some actual performance use case because these changes need thorough review, which require a lot of brain bandwidth. I have limited faith in end-to-end testing currently for aliasing given flang has a rather limited usage of it, so my concern is the introduction of hard to catch dormant optimization bugs (so thank you for all the LIT tests you are adding in your patch, this is great).

jdenny-ornl commented:
> Thanks @jdenny-ornl for the update. This looks better!

Thanks for reviewing.

> I am still a bit bothered by the fact that the code is turned in a way were it is proving aliasing instead of no aliasing (i.e., it is doing `if (cdt1) then mayalias else if (cdt2) then mayalias .... else noalias` ), this makes it harder to validate that aliasing cases are not missing via human review (at least in my experience). But I have to concede the standard is a bit written in that way in section 15.5.2.14 of F2023, and I do not have immediate suggestions to reverse the logic in a clearer way.

I think that is a description of [fir::AliasAnalysis::alias](https://github.com/llvm/llvm-project/blob/3b703d479ff37883242acc20fed317ed8a5466dc/flang/lib/Optimizer/Analysis/AliasAnalysis.cpp#L97-L181), and it holds with or without this PR.  I agree the reverse seems safer, but I don't know how to achieve it either.

> Can I ask you to push the `isRecordWithPointerComponent` and `isPointerReference` NFC change as well as the alias-analysis-9.fir renaming in a preliminary patch so that your functional changes stands out?

Done in PR #105899.  I haven't bothered with a stacked PR as I assume that will land quickly, but let me know if that would be helpful.

> Then I will run this update through some end-to-end apps and keep you posted.

Sounds great.  Thanks.

> Do you know any apps where I should see performance improvement

No, not so far.

> (i.e., was your change motivated by some use case in some app?).

Yes, indirectly.  I have developed some Flang alias analysis improvements that benefit the performance of LLNL's UMT proxy app. However, those improvements alter some alias analysis results in Flang's test suite.  Often where the existing expected result is MayAlias, I was originally not sure whether authors felt that is a necessary result or just a conservative result due to limitations of the current implementation.  I also then doubted new tests I was adding.  I wrote the RFC to get answers so I could make further progress on my improvements.  Developing this PR seemed like the best way to make the RFC discussion concrete.

> Note that in general I think we should try to change aliasing code in the future either to simplify the logic, or when motivated by some actual performance use case because these changes need thorough review, which require a lot of brain bandwidth. I have limited faith in end-to-end testing currently for aliasing given flang has a rather limited usage of it, so my concern is the introduction of hard to catch dormant optimization bugs (so thank you for all the LIT tests you are adding in your patch, this is great).

I understand.  When I started the RFC, I wasn't sure whether we'd end up landing changes to the implementation or just the test suite comments.  I hope the PR we ended up with serves to clarify both even as it exposes more detailed use cases in the Fortran standard.

jdenny-ornl committed:
Merge branch 'main' into flang-aa-ptr-component

jdenny-ornl committed:
Merge branch 'main' into flang-aa-ptr-component

jdenny-ornl commented:
> > Can I ask you to push the `isRecordWithPointerComponent` and `isPointerReference` NFC change as well as the alias-analysis-9.fir renaming in a preliminary patch so that your functional changes stands out?
> 
> Done in PR #105899. I haven't bothered with a stacked PR as I assume that will land quickly, but let me know if that would be helpful.

That has landed, and I've merged it here.

jdenny-ornl committed:
Remove todo about merging tests

jdenny-ornl committed:
Merge branch 'main' into flang-aa-ptr-component

jdenny-ornl commented on flang/test/Analysis/AliasAnalysis/ptr-component.fir:
```diff
@@ -0,0 +1,841 @@
+// is via fir.coordinate_of instead of hlfir.designate.  We would like alias
+// analysis results to be the same in both versions.
+//
+// TODO: What about merging with alias-analysis-3.fir as it has the same focus?
```
I removed the todo.

jdenny-ornl commented:
> Then I will run this update through some end-to-end apps and keep you posted.

@jeanPerier Any update here?  Let me know if there's any way I can help.

jdenny-ornl committed:
Merge branch 'main' into flang-aa-ptr-component

jdenny-ornl commented:
Ping

jdenny-ornl committed:
Merge branch 'main' into flang-aa-ptr-component

jdenny-ornl commented:
Ping.

jeanPerier commented on flang/include/flang/Optimizer/Analysis/AliasAnalysis.h:
```diff
@@ -157,6 +157,17 @@ struct AliasAnalysis {
+    /// @{
+    bool aliasesLikeDummyArg() const;
+    bool aliasesLikePtrDummyArg() const;
+    bool canBeActualArg() const;
+    bool canBeActualArgWithPtr(const mlir::Value *val) const;
```
Please add some short explanation, it is not very clear what is meant looking at the names it could be read as implying that they can be passed as actual argument in call inside the current function, not that they can be actual argument of the current function.

jdenny-ornl replied:
Done.

jeanPerier reviewed (commented):
Thanks for the updates @jdenny-ornl. Testing on a few apps seemed OK. Re-reading the PR, I was unconfutable with the names of the added helpers, so I am suggesting new names.

jdenny-ornl committed:
Merge branch 'main' into flang-aa-ptr-component

jdenny-ornl committed:
Improve some function names and comments

jdenny-ornl commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -94,6 +94,39 @@ bool AliasAnalysis::Source::isBoxData() const {
          origin.isData;
 }
 
+bool AliasAnalysis::Source::aliasesLikeDummyArg() const {
```
Done, but I went with `mayBeDummyArgOrHostAssoc` to make it a little shorter.  I think it's still clear given that we're talking about `SourceKind::HostAssoc`.

jdenny-ornl commented:
> Thanks for the updates @jdenny-ornl. Testing on a few apps seemed OK.

Good to hear.  Thanks.

jeanPerier reviewed (approved):
Thanks for the updates

jdenny-ornl commented:
Thanks for the review.
